### PR TITLE
Fix for - Downloader plugins not used when downloading new repo's ind…

### LINKED ISSUE #3938: Downloader plugins not used when downloading new repo's index.yaml
I've been attempting to use the [viglesiasce/helm-gcs](https://github.com/viglesiasce/helm-gcs) plugin to manage a private chart repo on Google Cloud Storage. That plugin implements a [very simple downloader](https://github.com/viglesiasce/helm-gcs/blob/master/bin/gcs-repo.sh) for the `gs://` protocol. From what I can tell, this is implemented correctly per the Helm docs on [downloader plugins](https://docs.helm.sh/using_helm/#downloader-plugins).

When I run: `helm repo add gcs-repo gs://the-dyrt-helm-charts`

It fails with: `Error: Looks like "gs://the-dyrt-helm-charts" is not a valid chart repository or cannot be reached: Get gs://the-dyrt-helm-charts/index.yaml: unsupported protocol scheme "gs"`

It seems like Helm is finding the custom `gs://` downloader, since other invalid protocol schemes fail with a different message: `Error: Could not find protocol handler for foo`

Digging around in the Helm source, I think the error is coming from trying to use an HTTP getter instead of the custom getter when fetching the index here: 
https://github.com/kubernetes/helm/blob/1a55161a53345816d724e1df9664f789113e1328/pkg/repo/chartrepo.go#L122

It looks like this always-HTTP behavior was introduced in #3206 when authentication support was added for remote repos. Before that change, the index download was done using the repo's instantiated client as expected.

---

Output of `helm version`:

```
Client: &version.Version{SemVer:"v2.9.0-rc4", GitCommit:"030d2c066bcf5a37931e0169b7921f3758d97095", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.9.0-rc4", GitCommit:"030d2c066bcf5a37931e0169b7921f3758d97095", GitTreeState:"clean"}
```

* Also tested on v2.9.0-rc3 and v2.8.2.

Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.6", GitCommit:"9f8ebd171479bec0ada837d7ee641dec2f8c6dd1", GitTreeState:"clean", BuildDate:"2018-03-21T15:21:50Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9+", GitVersion:"v1.9.6-gke.0", GitCommit:"cb151369f60073317da686a6ce7de36abe2bda8d", GitTreeState:"clean", BuildDate:"2018-03-21T19:01:20Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 

GKE

---

Mentioning https://github.com/viglesiasce/helm-gcs/issues/3 for a cross-issue link.

### PR DESCRIPTION
**Pull request by eyalbe4**
This PR fixes https://github.com/kubernetes/helm/issues/3938.

### CODE REVIEW

File: pkg/repo/chartrepo.go [hunk 1]
```diff
@@ -152,6 +150,14 @@ func (r *ChartRepository) DownloadIndexFile(cachePath string) error {
 	return ioutil.WriteFile(cp, index, 0644)
 }
 
+// If HttpGetter is used, this method sets the configured repository credentials on the HttpGetter.
+func (r *ChartRepository) setCredentials() {
+	var t *getter.HttpGetter
+	if reflect.TypeOf(r.Client) == reflect.TypeOf(t) {
```
Comment (adamreese) [inline pkg/repo/chartrepo.go]: Is there a reason for using reflect here?  You can cast the type:
```
	if t, ok := r.Client.(*getter.HttpGetter); ok {
		t.SetCredentials(r.Config.Username, r.Config.Password)
	}
```

File: pkg/downloader/chart_downloader.go [hunk 2]
```diff
@@ -180,80 +167,86 @@ func (c *ChartDownloader) ResolveChartVersionAndGetRepo(ref, version string) (*u
...
 	}
 
-	return u, r, g, nil
+	return u, r.Client, nil
+}
+
+// If HttpGetter is used, this method sets the configured repository credentials on the HttpGetter.
+func (c *ChartDownloader) setCredentials(r *repo.ChartRepository) {
+	var t *getter.HttpGetter
+	if reflect.TypeOf(r.Client) == reflect.TypeOf(t) {
```
Comment (adamreese) [inline pkg/downloader/chart_downloader.go]:
```
	if t, ok := r.Client.(*getter.HttpGetter); ok {
		t.SetCredentials(r.Config.Username, r.Config.Password)
	}
```

**Review by adamreese — CHANGES_REQUESTED** [review]

**Comment by bacongobbler** [comment]
Adam and I paired together on this one. The PR looks good to go. We found some enhancements that can be made (as @adamreese pointed out above). One those are fixed we can merge this.

@eyalbe4 do you have time to make those changes by EOD? If not then we can make another PR building upon this with those comments. This is the last PR to make it in before we cut the (hopefully) final release candidate for Helm 2.9. Thanks!

**Comment by bacongobbler** [comment]
Hey, I went ahead and made this changes on top of this PR in https://github.com/kubernetes/helm/pull/3968. Closing!

**Pull request closed without merging**